server/src/main/java/org/elasticsearch/action/support/NodeResponseTracker.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.support;
+
+import java.util.Collection;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReferenceArray;
+
+/**
+ * This class tracks the intermediate responses that will be used to create aggregated cluster response to a request. It also gives the
+ * possibility to discard the intermediate results when asked, for example when the initial request is cancelled, in order to release the
+ * resources.
+ */
+public class NodeResponseTracker {
+
+    private final AtomicInteger counter = new AtomicInteger();
+    private final AtomicReferenceArray<Boolean> receivedResponseFromNode;
+    private volatile AtomicReferenceArray<Object> responses;
```

Comment:
I think we can rely on responses never being `null` (i.e. we can assert this in `trackResponseAndCheckIfLast`, and reject nulls with a NPE too) so there's no need for a separate array of bools.

Reply:
I added this in order to be able to track which node hasn't responded yet after the responses have been discarded since discarding a response is assigning `null` to the array. Another way I can think of approaching in order to have only one array would be to I introduce something that symbolizes a discarded response and use that. What do you think about that approach?

Reply:
Sorry if I'm missing something here - my thinking is that we could distinguish these things with a single field: `responses == null` means we're discarding results, whereas if `responses != null` then we're still collecting results and we can use `responses.compareAndSet(i, null, response)` to capture the response from node `i` without the risk of double-counting.

Reply:
Oh, no maybe I am missing something here, so I think the discussion is good... 

So, the initial code was only "completing" the listener when all the nodes have responded. If we want to preserve this functionality, we still need to track which nodes haven't responded yet even if we get a cancellation, right? 

If we have `responses == null` we can keep counting the times that the `trackResponseAndCheckIfLast(..)` was called but we cannot detect a double call anymore. Maybe that's an acceptable risk or maybe completing the listener after all nodes have responded wasn't a feature but a result of the way the code was structured. I was a bit conservative changing this because I cannot fully guess all the side effects of such a change. 

Is my concern more clear now?

Reply:
Ah I see. I think we should ignore the risk of a double-call here given the extra complexity it needs to protect against it and the fact that I don't think it actually happens. IMO it should be safe enough just to assert there are no double-calls before cancellation.

Reply:
That's wonderful, I also had my doubts so since you also lean towards this, I will document it and clean up the code!